ansible/library/device_filter.py

```diff
@@ -105,23 +122,40 @@ def run_module():
     )
 
     device_name_filter = module.params['device_name_filter']
+    device_name_exclude_filter = module.params['device_name_exclude_filter']
     device_host_filter = module.params['device_host_filter']
+    device_host_exclude_filter = module.params['device_host_exclude_filter']
     device_model_filter = module.params['device_model_filter']
+    device_model_exclude_filter = module.params['device_model_exclude_filter']
     device_vendor_filter = module.params['device_vendor_filter']
+    device_vendor_exclude_filter = module.params['device_vendor_exclude_filter']
     device_size_filter = module.params['device_size_filter']
 
     try:
         for key, device in module.params['device_data'].items():
             print(key, device)
             if device_name_filter and not re.match(device_name_filter, key):
                 continue
-            if device_host_filter and not re.match(device_host_filter, device.get('host', '')):
-                continue
-            if device_model_filter and not re.match(device_model_filter, device.get('model', '')):
-                continue
-            if device_vendor_filter and not re.match(device_vendor_filter,
-                                                     device.get('vendor', '')):
+            if device_name_exclude_filter and re.match(device_name_exclude_filter, key):
                 continue
+            if len(device_host_filter):
```

Comment:
Are all these `if len(....)` needed? Could you just check if it is in the dict or not?

Reply:
The entire set of options is always sent to the device_filter.py module, so we do need to check list length on those options where we want to include the device based on the option.  For the exclusion options, the length check not necessary and have been removed.